src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let updatedState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      updatedState = { ...updatedState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => delete updatedState[key]);
+    } else if (action.type === 'clear') {
+      updatedState = {};
```

Comment:
The 'clear' action resets the state to an empty object. Ensure this behavior is intended and documented, as it completely removes all properties from the state.